endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Arguments.kt

```diff
@@ -0,0 +1,173 @@
+        return cliElement
+    }
+}
+
+/**
+ * Argument instance.
+ */
+abstract class Argument<TResult> internal constructor(owner: CLIEntityWrapper): CLIEntity<TResult>(owner)
+
+/**
+ * Common single argument instance.
+ */
+abstract class AbstractSingleArgument<T: Any, TResult> internal constructor(owner: CLIEntityWrapper): Argument<TResult>(owner) {
+    /**
+     * Check descriptor for this kind of argument.
+     */
+    internal fun checkDescriptor(descriptor: ArgDescriptor<*, *>) {
+        if (descriptor.number == null || descriptor.number > 1) {
+            error("Argument with single value can't be initialized with descriptor for multiple values.")
+        }
+    }
+}
+
+/**
+ * Argument with single non-nullable value.
+ */
+class SingleArgument<T : Any> internal constructor(descriptor: ArgDescriptor<T, T>, owner: CLIEntityWrapper):
+        AbstractSingleArgument<T, T>(owner) {
+    init {
+        checkDescriptor(descriptor)
+        cliElement = ArgumentSingleValue(descriptor)
+    }
+}
+
+/**
+ * Argument with single nullable value.
+ */
+class SingleNullableArgument<T : Any> internal constructor(descriptor: ArgDescriptor<T, T>, owner: CLIEntityWrapper):
+        AbstractSingleArgument<T, T?>(owner){
+    init {
+        checkDescriptor(descriptor)
+        cliElement = ArgumentSingleNullableValue(descriptor)
+    }
+}
+
+/**
+ * Argument with multiple values.
+ */
+class MultipleArgument<T : Any> internal constructor(descriptor: ArgDescriptor<T, List<T>>, owner: CLIEntityWrapper):
+        Argument<List<T>>(owner) {
+    init {
+        if (descriptor.number != null && descriptor.number == 1) {
+            error("Argument with multiple values can't be initialized with descriptor for single one.")
+        }
+        cliElement = ArgumentMultipleValues(descriptor)
+    }
+}
+
+/**
+ * Allow argument have several values.
+ *
+ * @param number number of arguments are expected. In case of null value any number of arguments can be set.
+ */
+fun <T : Any, TResult> AbstractSingleArgument<T, TResult>.number(value: Int? = null): MultipleArgument<T> {
+    if (value != null && value == 1) {
+        error("number() modifier with value 1 is unavailable. It's already set to 1.")
+    }
+    val newArgument = with((cliElement as ParsingValue<T, T>).descriptor as ArgDescriptor) {
+        MultipleArgument(ArgDescriptor(type, fullName, value, description, listOfNotNull(defaultValue),
+                required, deprecatedWarning), owner)
+    }
+    owner.entity = newArgument
+    return newArgument
+}
+
+/**
+ * Set default value for argument.
+ *
+ * @param value default value.
+ */
+fun <T: Any, TResult> AbstractSingleArgument<T, TResult>.default(value: T): SingleArgument<T> {
+    val newArgument = with((cliElement as ParsingValue<T, T>).descriptor as ArgDescriptor) {
+        if (required) {
+            printWarning("You can use optional(), because option with default value is defined.")
```

Comment:
So to avoid this warning, a user always has to write `.optional().default(...)`?

Reply:
Fixed